### PR TITLE
feat: XYNE-63875 self-tag — allow mentioning yourself in chat

### PR DESCRIPTION
**Pull request by XyneSpaces**
1. Problem Description:
Users cannot @-mention themselves in a channel or thread. The mention picker filters the current user out everywhere except a self-DM, so there is no way to tag yourself on a message you want to come back to. People work around it by DM-ing themselves or bookmarking.

2. If this is a bug fix or an enhancement, how did you identify the issue?:
Enhancement requested in a Spaces thread. Traced the mention picker (`packages/shared/src/hooks/useMentionSearch.ts`) and the message side-effect that turns mentions into activity/notifications (`apps/backend/src/zero/side-effects/tables/messages-handler.ts`).

3. Explain the solution implemented in the PR:
- Frontend: `useMentionSearch` now defaults `excludeSelf` to `false`, so you appear in the @-picker as `<name> (you)` in channels and threads. Callers that should not offer self can still pass `excludeSelf: true`. Self-DM behaviour is unchanged.
- Backend: a self-mention now records a `mentioned_user` activity for the sender, so the message shows up under "Your mentions" / Activity. The sender is deliberately kept OUT of `validMentionedUsers`, so a self-tag never produces a push/email notification, and never emits a `USER_MENTIONED` app event for your own message.

4. Documentation (link to docs created/updated for this change): N/A

5. DB Alter Details (if any): N/A — no schema change; reuses the existing `activities` row shape (`actorAction = 'mentioned_user'`, `userId = actorId = sender`).

6. Data fix Details (if any): N/A

7. Environment variable Changes (if any): N/A

8. Testing (how it was tested; screenshots/links):
Verified end-to-end in a sandbox dev environment (backend in test mode + dashboard):
- Typing `@` in a channel composer now lists `Test User (you)`.
- Selecting it inserts a normal mention chip and sends the message with the mention rendered.
- The Activity feed shows the resulting mention entry and the Activity badge increments.
- No push/email notification is generated for the self-mention (sender excluded from the notification path).
- `tsc --noEmit` on `apps/backend` passes; `@xyne/shared` builds clean.
A screen recording of the flow is attached in the originating Spaces thread.

9. Services to which this change is to be deployed: backend, dashboard (shared package rebuild).

10. Main PR link (if this PR is raised against a release branch): N/A

### CODE REVIEW

File: packages/shared/src/hooks/useMentionSearch.ts [hunk 1]
```diff
@@ -153,7 +153,9 @@ export const useMentionSearch = (
   threadParticipantIds?: ReadonlySet<string>,
   options: UseMentionSearchOptions = {},
 ): UseMentionSearchResult => {
-  const { includeSpecialMentions = true, excludeSelf = true } = options;
+  // Self-tag: you are a valid mention candidate everywhere. Callers can still
+  // opt out with excludeSelf: true (e.g. surfaces where tagging yourself is meaningless).
+  const { includeSpecialMentions = true, excludeSelf = false } = options;
```
Comment (XyneSpaces) [inline packages/shared/src/hooks/useMentionSearch.ts]: **[major] R2/R9 — flipping the shared default changes every consumer that doesn't pass the flag.** Code search shows `excludeSelf` is passed by exactly one caller — `SendMessageRichTextField.tsx` (automation composer), whose behaviour is unaffected by a default change. Every other consumer of this hook (ChatInput, ComposeDmPanel, QuickDmComposer, ForwardMessageModal, EntityShareModal, ScheduledMessageModal, XyneAIInputBox, ChannelCommandMenu, CanvasCommentsPanel, CanvasInlineCommentThread, …) silently flips from "self only in a self-DM" to "self everywhere". The PR scopes the feature to chat, and per the implementation handoff only the Canvas surfaces were audited.

Suggested fix: keep the default `excludeSelf = true` and opt in explicitly where self-mention is intended (pass `excludeSelf: false` from the chat composer), or audit/sign off on each affected surface in this PR.

Related R5: there is a dashboard-local copy of this hook at `apps/dashboard/src/hooks/useMentionSearch.ts` that doesn't carry the `excludeSelf` option at all (per code search) — the two hooks now diverge further. Assumption: the local copy is legacy and unaffected surfaces don't import it — please verify which surfaces import which copy.

File: apps/backend/src/zero/side-effects/tables/messages-handler.ts [hunk 2]
```diff
@@ -575,6 +575,14 @@ export class MessagesSideEffectHandler extends BaseSideEffectHandler {
         mentionSource: u.mentionSource
       }))
 
+    // Self-tag: tagging yourself records an activity so the message is findable
+    // under "Your mentions", but it is deliberately kept OUT of validMentionedUsers
+    // so it never produces a push/email notification or an app (USER_MENTIONED) event
+    // for your own message.
+    const isSelfMention = mentionedUsers.some(
```
Comment (XyneSpaces) [inline apps/backend/src/zero/side-effects/tables/messages-handler.ts]: **[major] R7 (review guideline 204 🟠) — new notification-path behaviour with no automated tests.** The self-tag split (sender gets a `mentioned_user` activity but no push/email/Slack/`USER_MENTIONED` event) is a behaviour change in the message side-effect path, verified only manually (POT recording in the originating thread).

Reading the full handler, the invariants do hold — `validMentionedUsers` and `notificationUserIds` both filter `userId !== senderId`, so the `activityRecipients` append can't duplicate and the sender can't be notified — but nothing pins them.

Suggested tests: (a) self-tag in a channel → exactly one activity for the sender (`actorAction: 'mentioned_user'`) and no notification; (b) mixed self + other-user mention → activities for both, notification only for the other user; (c) `@channel`/`@here` (group source) alone does NOT set `isSelfMention`; (d) other-user mention only → unchanged flow.

**Review by XyneSpaces — COMMENTED** [review]
Automated review by **Xyne Spaces Reviewer** — 2 major, 2 nits, no blocking issues. Full findings table and reviewer recommendation in the summary comment below.

**Comment by XyneSpaces** [comment]
## Xyne Spaces Reviewer — PR #1986

**Verdict:** Changes requested (2 major, 2 nits — no blocking issues)
**Reviewed:** 2 files, 4 hunks (all) @ head `cf24ace` · **Findings:** 0 blocking · 2 major · 2 nits

### Findings
| Severity | Rule | Location | Issue |
|----------|------|----------|-------|
| major | R2/R9 | `packages/shared/src/hooks/useMentionSearch.ts:158` | `excludeSelf` default flip silently changes ~15 consumers that don't pass the flag (only Canvas surfaces audited); divergent dashboard-local hook copy (R5) |
| major | R7 / guideline 204 | `apps/backend/src/zero/side-effects/tables/messages-handler.ts:582` | new notification-path behaviour with no automated tests |
| nit | guideline 294 | PR title / branch | missing XYNE ticket id — team convention is `feat: XYNE-NNNNN <subject>` (cf. #1542, #1478) |
| nit | — | out of diff (ActivityItem copy) | a self-tag activity will render as "You mentioned you" — known open copy decision, flagging so it's tracked |

**Verified while reviewing (no finding):** the PR's core backend invariant holds — `validMentionedUsers` and `notificationUserIds` both pre-filter `userId !== senderId`, so a self-tag records exactly one `mentioned_user` activity and can never produce a duplicate activity or a push/email/app notification. The backend change is correct; it just isn't pinned by tests (finding 2).

### Recommended reviewer
**@anuragdvd** (Anurag Dwivedi · anurag.dwivedi@juspay.in) — owner of the `useMentionSearch` hook this PR modifies; the frontend default-flip is the PR's main risk and he knows the call sites. Zero pending review requests in this repo.

| Candidate | Owns (files in this PR) | CODEOWNER match? | Pending reviews | Tie-breaker |
|-----------|-------------------------|------------------|-----------------|--------------|
| Anurag Dwivedi (@anuragdvd) | `useMentionSearch.ts` (hook owner per Spaces history) | no | 0 (open-PR list) | active in Spaces, last seen 12 Sep 2026 |
| Prakhar Agarwal | `mentioned_user` activity path (backend) | no | — | **excluded: INACTIVE in Spaces since 28 Apr 2026** |
| Utkarsh Pandey | tagging popup UX | no | not computed (GitHub login unresolved) | — |
| Radheyshree Agrawal | notifications | no | not computed (GitHub login unresolved) | — |

Neither changed file matches a `.github/CODEOWNERS` path (CODEOWNERS covers ACL/mutator dirs only), so the ranking is file-ownership + capacity. Workload = appearances as requested reviewer across the repo's open PRs; `@anuragdvd` appears in none.

⚠️ This automation run has no GitHub reviewer-request capability (no request-reviewers tool granted, no `gh` CLI fallback), so **@anuragdvd could not be formally added as a reviewer via the API** — please add him on the PR manually, or Anurag, feel free to pick this up from this mention.

**Comment by XyneSpaces** [comment]
Review feedback addressed in e11795dc.

- **[major] shared default flip** — reverted; `excludeSelf` defaults to `true` again and self-tag is opted into only by the chat composer (`ChatInput.tsx`). No other mention surface changes behaviour. The dashboard hook is a forwarding wrapper, not a divergent copy.
- **[major] no automated tests** — the activity-vs-notification split is now in `selfMention.ts` with 6 unit tests in `selfMention.test.ts` covering all four scenarios you listed plus two edge cases.
- **[nit] activity copy** — a self-tag now renders "You tagged yourself in #channel" instead of "<name> mentioned you in"; handled in `MessageMentionActivity.tsx` via `activity.actorId === activity.userId`.
- **[nit] missing XYNE ticket id** — no ticket key exists yet for this work; the branch and title will be renamed to `feat: XYNE-NNNNN …` once the ticket is filed.

Re-verified end-to-end in a sandbox dev environment after the changes: the picker still lists `Test User (you)` in the channel composer, and the Activity feed shows the new self-tag copy.

**Comment by XyneSpaces** [comment]
Fixed in e11795dc. Pulled the split into two pure functions in the new `apps/backend/src/zero/side-effects/tables/selfMention.ts` (`isSelfMention`, `buildMentionActivityRecipients`) and pinned the invariants in `selfMention.test.ts` — 6 tests, all passing:

- self-tag in a channel: exactly one `mentioned_user` activity for the sender, sender absent from the notification/app-event set
- self + other mention: activities for both, notification only for the other user
- `@channel`/`@here` (group source) alone does not set `isSelfMention`
- other-user mention only: flow unchanged
- self-tag in a channel the sender does not participate in is ignored
- appending the sender can never duplicate an existing activity

Verified: `npx jest selfMention.test.ts` → 6 passed; `tsc --noEmit` clean on apps/backend, apps/dashboard and packages/shared.

**Comment by XyneSpaces** [comment]
Fixed in e11795dc. `excludeSelf` now defaults back to `true` in the shared hook, and the chat composer opts in explicitly — `ChatInput.tsx` passes `excludeSelf: false`. Every other consumer (ForwardMessageModal, ComposeDmPanel, QuickDmComposer, ScheduledMessageModal, XyneAIInputBox, ChannelCommandMenu, EntityShareModal, CanvasCommentsPanel, CanvasInlineCommentThread, SendMessageRichTextField) is byte-for-byte unchanged in behaviour.

On R5: `apps/dashboard/src/hooks/useMentionSearch.ts` is a thin wrapper, not a divergent copy — it injects `user?.id` from `useAuth()` and forwards the `options` object (typed `UseMentionSearchOptions`) straight through to the shared hook, so `excludeSelf` does reach it. All dashboard surfaces import that wrapper; nothing imports the shared hook directly.

File: apps/backend/src/zero/side-effects/tables/messages-handler.ts [hunk 3]
```diff
@@ -575,6 +576,13 @@ export class MessagesSideEffectHandler extends BaseSideEffectHandler {
         mentionSource: u.mentionSource
       }))
 
+    // Self-tag: tagging yourself records an activity so the message is findable
+    // under "Your mentions", but the sender is deliberately kept OUT of
+    // validMentionedUsers (and therefore out of notificationUserIds and the
+    // USER_MENTIONED app event), so you are never notified about your own
+    // message. The split is unit-tested in ./selfMention.test.ts.
+    const selfMention = isSelfMention(mentionedUsers, senderId, channelParticipantIds);
```
Comment (XyneSpaces) [inline apps/backend/src/zero/side-effects/tables/messages-handler.ts:581]: **[minor] R2** — Coverage gap (assumption-labeled): this `selfMention` activity path covers channel messages, channel thread replies, and DMs, but the group-DM *thread-reply* branch of `upsertReplyActivityV2` still builds `groupDMMentionedUsers` with its own `u.userId !== senderId` predicate and its own `createMentionActivities` call — a self-tag in a group-DM thread reply would record no activity. I could not fully confirm from this diff whether group-DM replies also flow through this main path (the branch is outside the PR's hunks); if they don't, consider routing that branch through `isSelfMention`/`buildMentionActivityRecipients` too (or a follow-up ticket) so self-tag behaves uniformly across surfaces.

File: apps/backend/src/zero/side-effects/tables/selfMention.test.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+const SENDER = 'user-sender';
+const OTHER = 'user-other';
+const PARTICIPANTS = new Set([SENDER, OTHER]);
+
+/**
+ * Mirrors the handler: `validMentionedUsers` keeps only direct/group mentions of
+ * channel participants, and always drops the sender. Notifications are derived
+ * from the same sender-excluded set.
+ */
+function buildValidMentionedUsers(
```
Comment (XyneSpaces) [inline apps/backend/src/zero/side-effects/tables/selfMention.test.ts]: **[nit] R7** — `buildValidMentionedUsers` here mirrors the handler's production filter (mentionSource whitelist + `participantSet.has` + `userId !== senderId`) instead of pinning it: if the production predicate drifts (e.g. a new `mentionSource` value), this copy keeps the old behaviour and the tests still pass. Consider exporting the real filter from the handler (or `selfMention.ts`) and testing against it, or adding a handler-level invariant test that APP_MENTION/USER_MENTIONED activities never include the sender.

File: apps/backend/src/zero/side-effects/tables/selfMention.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
+/**
+ * Self-tag (self-mention) helpers for the message side-effect path.
+ *
+ * Tagging yourself must record an activity (so the message is findable under
+ * "Your mentions") WITHOUT ever notifying you about your own message. The split
+ * lives here as pure functions so the invariant is unit-testable instead of
+ * being buried in the side-effect handler.
+ */
+
+export type MentionSource = 'direct' | 'group' | 'channel' | 'here';
```
Comment (XyneSpaces) [inline apps/backend/src/zero/side-effects/tables/selfMention.ts:2]: **[nit] R8** — `MentionSource` redeclares the union already carried by `ExtractedMentionForNotification['mentionSource']` in `apps/backend/src/utils/mentionUtils.ts` (`'direct' | 'group' | 'channel' | 'here'`). Two hand-maintained copies can drift when a new mention source is added. Consider exporting `MentionSource` from `mentionUtils.ts` and importing it here (and in the test) for a single source of truth.

**Review by XyneSpaces — COMMENTED** [review]
Re-review at `e11795dc` (new commits since my last review at `cf24ace9`). Checked every changed hunk (6 files, 10 hunks).

**All four previous findings are addressed** ✅
- ✅ (major 1) `excludeSelf` default restored to `true`; ChatInput is the only surface opting in (`excludeSelf: false`) — the other 10 consumers keep historical behaviour.
- ✅ (major 2) `isSelfMention` / `buildMentionActivityRecipients` extracted as pure functions in `selfMention.ts` with 6 unit tests.
- ✅ (nit 3) Self-mention copy fixed — "You tagged yourself in" via `activity.actorId === activity.userId`.
- ⚠️ (nit 4) Ticket id still missing from the title — **now blocking**: the required **Ticket Validation** check fails on this head (`PR title must format: PROJECT-####: description OR feat/fix: PROJECT-####: subject`). Retitle to e.g. `feat: XYNE-####: self-tag — allow mentioning yourself in chat` and it should go green.

One minor + two nits inline. Full summary and reviewer recommendation in the follow-up comment.

**Comment by XyneSpaces** [comment]
## Xyne Spaces Reviewer — PR #1986

**Verdict:** Blocking issues found (trivial fix: add ticket id to the PR title)
**Reviewed:** 6 files, 10 hunks @ `e11795dc` · **Findings:** 1 blocking, 1 minor, 3 nits

### Findings
| Severity | Rule | Location | Issue |
|----------|------|----------|-------|
| blocking | CI gate | PR title | Required **Ticket Validation** check fails on head `e11795dc` — title lacks a ticket id (must match `feat: XYNE-####: …`). Retitle to merge. |
| minor | R2 | apps/backend/src/zero/side-effects/tables/messages-handler.ts:584 | Group-DM thread replies don't route through the new self-tag activity path (assumption-labeled — see inline comment). |
| nit | R7 | apps/backend/src/zero/side-effects/tables/selfMention.test.ts:17 | Test mirrors the production filter; drift would go undetected. |
| nit | R8 | apps/backend/src/zero/side-effects/tables/selfMention.ts:10 | `MentionSource` union duplicated from `apps/backend/src/utils/mentionUtils.ts`. |
| nit | R9 | PR description | Description still documents the old design ("`useMentionSearch` now defaults `excludeSelf` to `false`") — the code now defaults `true` with ChatInput opting in. Update to match. |

**Previous round (`cf24ace9`):** all 4 findings addressed ✅ — default `excludeSelf = true` restored (ChatInput opt-in), pure `isSelfMention`/`buildMentionActivityRecipients` helpers + 6 unit tests, self-mention copy fixed. Only the ticket-id nit remains, escalated to blocking by the failing required CI check.

### Recommended reviewer
**@VishalM-23** — the only candidate with verified ownership of a file changed in this PR (`packages/shared/src/hooks/useMentionSearch.ts`, PR #933 +78/−45), owner of the merged mention fix XYNE-16227, and the freshest Spaces presence of the candidates.

| Candidate | Owns (files in this PR) | CODEOWNER? | Pending reviews | Tie-breaker |
|-----------|-------------------------|-----------|-----------------|-------------|
| @VishalM-23 (Vishal M) | `packages/shared/src/hooks/useMentionSearch.ts` (PR #933); mention-fix XYNE-16227 owner | no — repo CODEOWNERS only covers ACL paths | ~317 review-requested search hits (approximate, non-differentiating) | last Spaces activity 16 Sep — freshest |
| @anuragdvd (Anurag Dwivedi) | — (recommended last round as useMentionSearch owner, but no verified touch of this PR's 6 files) | no | ~375 review-requested search hits (approximate) | last Spaces activity 12 Sep |

Rationale: no CODEOWNER matches the changed paths (repo CODEOWNERS covers ACL paths only), so ranking is by file ownership + Spaces activity — @VishalM-23 wins on both.

*Note: this automation has no `request_reviewers` API access, so the reviewer is @mentioned here rather than formally added as a requested reviewer.*

File: apps/backend/src/zero/side-effects/tables/selfMention.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+export interface MentionedUser {
+  userId: string;
+  mentionSource: MentionSource;
+}
+
+export type MentionActivityRecipient = MentionedUser;
+
+// Only a direct `@self` by a channel participant counts — `@channel`/`@here`
+// expansion includes the sender and must not be treated as a self-tag.
+export function isSelfMention(
```
Comment (XyneSpaces) [inline apps/backend/src/zero/side-effects/tables/selfMention.ts:13]: **[major] R7 (tests)** — `8b368934` deletes `selfMention.test.ts` (90 lines, 6 tests), the only automated coverage of this module's two invariants: (1) `isSelfMention` counts only a *direct* `@self` by a channel participant — `@channel`/`@here` expansion must not be treated as a self-tag; (2) `buildMentionActivityRecipients` appends the sender only on a true self-tag and never when already present in `validMentionedUsers` — the invariant that keeps a self-mention out of push/email and `USER_MENTIONED` for your own message. This regresses the round-1 R7 fix that landed in `e11795dc`; a future change to either predicate (e.g. a new `mentionSource` value) would now pass CI silently. I understand the deletion was requested in the thread — if dropping the suite is a deliberate policy call, either restore it from `e11795dc` (`git checkout e11795dc -- apps/backend/src/zero/side-effects/tables/selfMention.test.ts`) and slim it to the two invariants, or record the decision in the PR description with a follow-up ticket so the suppression invariant has a tracked owner.

**Review by XyneSpaces — COMMENTED** [review]
Re-review at `8b368934` — new commits since my last review at `e11795dc`. Every changed hunk in `e11795dc..8b368934` checked (6 files, −134/+14: five comment trims + the deletion of `selfMention.test.ts`).

- Comment trims: fine — no behaviour change; the trimmed blocks restate what the code already expresses.
- One new major finding (R7) inline below: the deleted test suite was the only automated coverage of the notification-suppression invariant.
- No Prisma migration/schema paths touched — migration-safety gate not triggered.

Full round-3 summary + reviewer recommendation in the follow-up comment.

**Comment by XyneSpaces** [comment]
## Xyne Spaces Reviewer — PR #1986

**Verdict:** Changes requested — CI "Ticket Validation" failing (blocks merge) + test coverage removed (major).
**Reviewed:** round 3 at `8b368934` — new commits `e11795dc..8b368934`, 6 files, −134/+14 · **Findings this round:** 1 blocking (CI), 1 major · carried over: 1 minor, 1 nit

### Findings (this round)
| Severity | Rule | Location | Issue |
|----------|------|----------|-------|
| blocking | CI gate (carried from round 2) | PR title | "Ticket Validation" still fails at `8b368934` — title needs a Jira id, e.g. `feat: XYNE-####: self-tag — allow mentioning yourself in chat`. No XYNE ticket is linked in the description; the id has to come from the feature owner. |
| major | R7 (tests) | `apps/backend/src/zero/side-effects/tables/selfMention.ts:13` | `8b368934` deletes `selfMention.test.ts` — the only automated coverage of the notification-suppression invariant (self-mention → activity row; sender never enters `validMentionedUsers` → no push/email/`USER_MENTIONED` for your own message). Regresses the round-1 R7 fix from `e11795dc`. Inline comment posted. |

### Carried over (unchanged by these commits — not re-posted inline)
- **minor R2** — group-DM *thread replies*: the `groupDMMentionedUsers` branch of `upsertReplyActivityV2` in `messages-handler.ts` keeps its own `u.userId !== senderId` predicate, so a self-tag in a group-DM thread reply records no activity (assumption-labeled, round 1).
- **nit R8** — `MentionSource` (`selfMention.ts:2`) redeclares the union from `apps/backend/src/utils/mentionUtils.ts`; export from one place (round 1).
- The round-1 nit that the test mirrored the production filter is moot now the tests are deleted — but see the major above.

### Migration safety
No changed path matches `apps/backend/prisma/**` (or legacy `backend/prisma/**`) — gate not triggered, nothing to report.

### Recommended reviewer
**@VishalM-23** (Vishal M) — owns the mentions domain (assignee of XYNE-16227 "Fix @mentions not appearing in channel"), most recently active candidate in Spaces (16/9), and the reviewer already recommended in the feature thread.

| Candidate | Owns (files in this PR) | CODEOWNER? | Pending reviews | Tie-breaker |
|-----------|-------------------------|-----------|-----------------|-------------|
| @VishalM-23 | — (domain owner of the mentions path; XYNE-16227 assignee) | no | not computable¹ | Spaces-active 16/9; prior recommendation in thread |
| @anuragdvd | head commit `8b368934` (this round's changes) | no | not computable¹ | excluded as primary — authored the changes under review |
| @harshpreet931, @ameer030, @Revanthvenkatpasupuleti | recent commits in the touched area (not attributable per-file²) | no | not computable¹ | no confirmed Spaces mapping |

¹ The available search tool ignores the `review-requested:` qualifier (it falls back to lexical matching across all 317 open PRs — its top hit is a throwaway PR with no reviewers requested), so per-person pending-review counts cannot be computed; workload was judged from Spaces activity instead.
² Per-file ownership via commits-by-path is not available to this bot (the commits tool has no path filter), and `.github/CODEOWNERS` covers only ACL/mutator paths — none touched here. The ranking therefore rests on domain ownership + Spaces activity, stated as such.

⚠️ **Reviewer request:** the formal reviewer-request API is not available to this bot in this environment (no reviewer-request tool, no `gh` CLI), so @VishalM-23 is tagged here instead — someone with access please run `gh pr edit 1986 --add-reviewer VishalM-23` (or add him via the UI).

**Top blocker:** the "Ticket Validation" CI check — the PR title needs the `XYNE-####` ticket id from the feature owner.

**Review by Mohan0544 — APPROVED** [review]

**Review by Mohan0544 — DISMISSED** [review]

File: apps/backend/src/zero/side-effects/tables/selfMention.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  return [...validMentionedUsers, { userId: senderId, mentionSource: 'direct' }];
+}
+
+// Cheap, synchronous pre-filter for the DM / group-DM paths, which (unlike the
+// channel path) do not already extract mentions for every message. Returns true
+// only when the raw content *could* carry a direct @self, so the expensive
+// extraction runs on self-tag messages instead of on every DM.
+const ZWSP = '\u200B';
+
+export function mayContainSelfMention(content: string, senderId: string): boolean {
```
Comment (XyneSpaces) [inline apps/backend/src/zero/side-effects/tables/selfMention.ts]: **[minor] R5 (g263/g232)** — `mayContainSelfMention` independently re-encodes the exact set of mention formats that `extractAllUsersForNotification` → `extractMentionsFromContent` (mentionUtils.ts) parses: the FlowJSON mrkdwn token `<userid:ID>`, the `data-user-id` span, and the legacy ZWSP `@username`. I verified all three currently line up (the pre-filter is a superset for the span/ZWSP shapes, so there are no false negatives today) — but nothing links the two modules: a format added to the extractor, or a regex shape change there, would silently drop DM/group-DM self-tag activities while the channel path keeps working — no error, no log.

Suggested fix: export the format patterns from mentionUtils.ts (or a shared constants module) and consume them here, or add a cross-module sync test that feeds one fixture per extractor format through `mayContainSelfMention` so a new extractor format fails this pre-filter's suite.

File: apps/backend/src/zero/side-effects/tables/messages-handler.ts [hunk 8]
```diff
@@ -1484,6 +1495,52 @@ export class MessagesSideEffectHandler extends BaseSideEffectHandler {
...
+
+      // channelId is deliberately omitted: @channel/@here expansion is irrelevant
+      // to a self-tag and would cost an extra participant lookup.
+      const mentioned = await extractAllUsersForNotification(
+        contentForMentions,
+        this.ctx.workspaceId,
+      );
+      if (!isSelfMention(mentioned, senderId, participantIds)) return;
+
+      await activityService.createActivities([{
```
Comment (XyneSpaces) [inline apps/backend/src/zero/side-effects/tables/messages-handler.ts]: **[minor] R5 (g263)** — this hand-built activity literal duplicates the channel path's row construction (`activityRecipients.map(user => ({ ... }))`, ~line 667 in this file) field-for-field: `id` / `userId` / `actorId` / `actorAction` / `actionSource` / `actionSourceId` / `messageId` / `channelId` / `isThreadActivity` / `classification`. Two copies of the same row shape can drift — e.g. a future field added to channel mention activities but not to DM self-tag activities.

Suggested fix: extract a shared builder, e.g. `buildMentionActivityRows(recipients, { senderId, messageId, channelId, isThreadActivity })`, and use it from both `createSelfMentionActivityForDm` and the channel path.

File: apps/backend/src/zero/side-effects/tables/messages-handler.ts [hunk 9]
```diff
@@ -1508,6 +1565,18 @@ export class MessagesSideEffectHandler extends BaseSideEffectHandler {
       await this.handleSpecialMentionActivities(channelId, messageId, senderId, mentionType, [], initialMessageId !== messageId);
     }
 
+    // Self-tag parity with channels: tagging yourself in a DM or group DM records
+    // an activity (never a notification). Handled once here because the DM paths
+    // below build their mention lists per scope/thread and all exclude the sender.
+    await this.createSelfMentionActivityForDm(
```
Comment (XyneSpaces) [inline apps/backend/src/zero/side-effects/tables/messages-handler.ts]: **[nit] R7 (g210)** — the restored suite covers `isSelfMention` / `buildMentionActivityRecipients` / `mayContainSelfMention` as units, but nothing pins this wiring: DM path creating exactly one activity, group-DM thread-reply path (`isReply` → `isThreadActivity: true`), and pre-filter miss → no `extractAllUsersForNotification` call and no activity. Since parity with the channel path is this PR's core claim, a handler-level test (mock `extractAllUsersForNotification` + `activityService.createActivities`) would lock it in. Non-blocking — the units are well covered.

**Review by XyneSpaces — COMMENTED** [review]
**Round 4 re-review** — new hunks since round 3 (`8b368934`): `5cb068ec` adds the DM / group-DM self-tag activity path. Walked every new hunk (messages-handler.ts, selfMention.ts, selfMention.test.ts). 3 inline findings: 2 minor (R5 — duplicated mention-format knowledge between the pre-filter and the extractor; duplicated activity-row construction vs the channel path), 1 nit (R7 — no test for the handler wiring). No blocking or major issues.

Prior-round status: ticket id in title ✅ fixed (Ticket Validation passed on `5cb068ec`), deleted test suite ✅ restored (12 tests), group-DM thread-coverage question ✅ closed (single call before the scope branches, `isReply` → `isThreadActivity`). Full summary follows as a top-level comment.

**Comment by XyneSpaces** [comment]
## Xyne Spaces Reviewer — PR #1986 (round 4)

**Verdict:** Approve-worthy after nits
**Reviewed:** new hunks in `5cb068ec` — messages-handler.ts, selfMention.ts, selfMention.test.ts (3 files, 6 hunks) · **Findings:** 0 blocking, 0 major, 2 minor, 1 nit (+2 nits carried over from earlier rounds)

Round 4 re-review, triggered by new commits since round 3 (`8b368934`): merge `0f9eb4d1` + `5cb068ec` (DM/group-DM self-tag activity). The UI files (MessageMentionActivity.tsx, ChatInput.tsx, useMentionSearch.ts) are unchanged since round 3 and are not re-commented.

### Prior-round status
| Round | Finding | Status |
|-------|---------|--------|
| 2 | Ticket id missing from title (blocking) | ✅ fixed — title carries `XYNE-63875`; Ticket Validation passed on `5cb068ec` |
| 3 | Test suite deleted (R7 major) | ✅ fixed — `selfMention.test.ts` restored, 12 tests |
| 2 | Group-DM thread-reply coverage question (`upsertReplyActivityV2` branch) | ✅ closed — the single new call before the scope branches covers parent + thread reply via `isReply` |
| 1–2 | `MentionSource` union duplicated vs mentionUtils.ts (nit) | still open |
| 2 | test mirrors production filter via `buildValidMentionedUsers` (nit) | still open |

### Findings (new hunks only)
| Severity | Rule | Location | Issue |
|----------|------|----------|-------|
| minor | R5 | apps/backend/src/zero/side-effects/tables/selfMention.ts:41 | `mayContainSelfMention` re-encodes the extractor's three mention formats independently — future drift would silently drop DM self-tags while channels keep working (verified in sync today) |
| minor | R5 | apps/backend/src/zero/side-effects/tables/messages-handler.ts:1526 | activity literal duplicates the channel-path row construction (~line 667) field-for-field |
| nit | R7 | apps/backend/src/zero/side-effects/tables/messages-handler.ts:1571 | no test pins the DM/group-DM handler wiring (activity created, thread-reply path, pre-filter miss) |

Migration safety: not triggered — no `prisma/migrations/**` or `schema.prisma` paths in this PR's diff.

### Recommended reviewer
**@Mohan0544** (Mohan Kumar Malapureddy — mohankumar.malapureddy@juspay.in) — reviewed and approved this PR's diff at `0f9eb4d1` (approval later auto-dismissed by new pushes), is active in the #xyne-spaces PR-1986 thread, and was explicitly asked on 19/9 by the ticket creator to review it; he has 0 pending review requests across the repo's open PRs.

| Candidate | Familiarity with this PR | CODEOWNERS match? | Pending reviews | Tie-breaker |
|-----------|--------------------------|-------------------|-----------------|-------------|
| @Mohan0544 (Mohan Kumar Malapureddy) | reviewed + approved the full diff at `0f9eb4d1` | no — no CODEOWNERS entry covers these paths | 0 | prior approver; explicitly requested by the ticket creator |
| Naman Kalkhuria | created XYNE-63875, active in the PR thread | no | 0 | ticket creator; has not reviewed the code |
| Satvik Batra | flagged the DM gap; authored the head commit | no | — | excluded: requester / head-commit author (his approval is already the pending gate) |

Signals that could not be computed: per-path commit authorship (the available commits API has no path filter and no `gh` CLI exists in this environment — recent-main-committer list was used only to exclude bots), and CODEOWNERS coverage (repo `.github/CODEOWNERS` only covers acl/tenant/mutator paths, none of this PR's files). The reviewer-request API is not available in this agent's toolset, so @Mohan0544 could not be formally added as a requested reviewer — please add him via `gh pr edit 1986 --add-reviewer Mohan0544` or the PR sidebar.

Full details are in the 3 inline review comments on `5cb068ec`.

**Review by satvikbatra — APPROVED** [review]

**Review by Mohan0544 — APPROVED** [review]
